### PR TITLE
Add collection types for IR children

### PR DESCRIPTION
**Pull request by rynowak**
/cc @NTaylorMullen

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/ReadOnlyIRNodeCollection.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language.Intermediate
+{
+    public sealed class ReadOnlyIRNodeCollection : RazorIRNodeCollection
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/ReadOnlyIRNodeCollection.cs]: ReadOnly is kind of misleading here. It's more like `EmptyIRNodeCollection`

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/DefaultIRNodeCollection.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        }
+
+        public override bool Contains(RazorIRNode item)
+        {
+            return _inner.Contains(item);
+        }
+
+        public override void CopyTo(RazorIRNode[] array, int arrayIndex)
+        {
+            if (array == null)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/DefaultIRNodeCollection.cs]: Does the inner not do all these checks?

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/DefaultIRNodeCollection.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            return _inner.IndexOf(item);
+        }
+
+        public override void Insert(int index, RazorIRNode item)
+        {
+            if (index < 0 || index > Count)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/DefaultIRNodeCollection.cs]: ditto and elsewhere.

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/DefaultIRNodeCollection.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/DefaultIRNodeCollection.cs]: I'll get the opposite feedback from the next person to look at it.

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/ReadOnlyIRNodeCollection.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/ReadOnlyIRNodeCollection.cs]: I disagree. It's whole purpose is to be read only. It happens to also be empty.

**Comment by NTaylorMullen** [comment]
This be good.

**Pull request merged**